### PR TITLE
Transform CatEntry into AstroObject with minimal common indexing roun…

### PR DESCRIPTION
**Pull request by pirogronian**
…tines.
It's a small step for programmer, but big step for Celestia... Well, just joking. This is my desperate attempt to introduce at least some of AstroDB functionality without blowing everything up, including my mind. Just little steps, with continuous community support.
However, I can't promise how far I will proceed this process. I'm still unsure about my usability.
I couldn't test code for Win and Mac, I will only check Appveyor and Travis for errors.

### CODE REVIEW

File: src/celengine/astroobj.h [hunk 1]
```diff
@@ -7,9 +7,21 @@
 
 class UserCategory;
 
-class CatEntry
+class AstroCatalog
 {
+ public:
+    typedef unsigned int IndexNumber;
```
Comment (hv-work) [inline src/celengine/astroobj.h]: please `uint32_t` here

File: src/celengine/astroobj.h [hunk 2]
```diff
@@ -7,9 +7,21 @@
...
+    static const IndexNumber InvalidIndex = UINT_MAX;
```
Comment (hv-work) [inline src/celengine/astroobj.h]: and `UINT32_MAX` here

File: src/celestia/win32/winmain.cpp [hunk 3]
```diff
@@ -1673,7 +1673,7 @@ void ShowWWWInfo(const Selection& sel)
             if (url.empty())
             {
                 char name[32];
-                sprintf(name, "HIP%d", sel.star()->getCatalogNumber() & ~0xf0000000);
+                sprintf(name, "HIP%d", sel.star()->getIndex() & ~0xf0000000);
```
Comment (hv-work) [inline src/celestia/win32/winmain.cpp]: As you changed this line, could you replace `0xf0000000` with `Star::MaxTychoCatalogNumber`?

File: src/celengine/dsodb.cpp [hunk 4]
```diff
@@ -71,14 +71,14 @@ DSODatabase::~DSODatabase()
 DeepSkyObject* DSODatabase::find(const uint32_t catalogNumber) const
```
Comment (hv-work) [inline src/celengine/dsodb.cpp]: could you use `AstroCatalog::IndexNumber` instead of `uint32_t` here and in other cases?

**Comment by pirogronian** [comment]
I messed something up with commits, sorry. Just wait for next commit.

File: src/celengine/dsodb.cpp [hunk 5]
```diff
@@ -240,7 +240,7 @@ bool DSODatabase::load(istream& in, const fs::path& resourcePath)
         objType = tokenizer.getNameValue();
 
         bool   autoGenCatalogNumber   = true;
-        uint32_t objCatalogNumber     = DeepSkyObject::InvalidCatalogNumber;
+        uint32_t objCatalogNumber     = AstroCatalog::InvalidIndex;
```
Comment (375gnu) [inline src/celengine/dsodb.cpp]: you missed uint32_t here

File: src/celengine/selection.cpp [hunk 6]
```diff
@@ -116,12 +116,12 @@ string Selection::getName(bool i18n) const
     {
     case Type_Star:
         {
-            return fmt::sprintf("#%d", star()->getCatalogNumber());
+            return fmt::sprintf("#%d", star()->getIndex());
```
Comment (375gnu) [inline src/celengine/selection.cpp]: %d is signed, please change to %u

File: src/celengine/selection.cpp [hunk 7]
```diff
@@ -116,12 +116,12 @@ string Selection::getName(bool i18n) const
...
         }
 
     case Type_DeepSky:
         {
-            return fmt::sprintf("#%d", deepsky()->getCatalogNumber());
+            return fmt::sprintf("#%d", deepsky()->getIndex());
```
Comment (375gnu) [inline src/celengine/selection.cpp]: as above.

I also doubt that we need curly braces here, could you remove them?

File: src/celengine/selection.cpp [hunk 8]
```diff
@@ -142,7 +142,7 @@ string Selection::getName(bool i18n) const
                     if (parentStar != nullptr)
                     {
                         string buf;
-                        buf = fmt::sprintf("#%d", parentStar->getCatalogNumber());
+                        buf = fmt::sprintf("#%d", parentStar->getIndex());
                         name = buf + '/' + name;
```
Comment (375gnu) [inline src/celengine/selection.cpp]: as above regarding %d, and the whole block can be simplified to
```
name = fmt::sprintf("#%u/%s", parentStar->getIndex(), name);
```

File: src/celengine/stardb.cpp [hunk 9]
```diff
@@ -267,11 +267,11 @@ Star* StarDatabase::find(const string& name) const
 uint32_t StarDatabase::crossIndex(const Catalog catalog, const uint32_t celCatalogNumber) const
 {
     if (static_cast<uint32_t>(catalog) >= crossIndexes.size())
```
Comment (375gnu) [inline src/celengine/stardb.cpp]: i suppose crossIndexes.size() is size_t, so `static_cast<size_t>(catalog)`

File: src/celestia/qt/qtcelestialbrowser.cpp [hunk 10]
```diff
@@ -191,9 +191,9 @@ QVariant StarTableModel::data(const QModelIndex& index, int role) const
         {
         case NameColumn:
             {
-                uint32_t hipCatNo = star->getCatalogNumber();
+                uint32_t hipCatNo = star->getIndex();
```
Comment (375gnu) [inline src/celestia/qt/qtcelestialbrowser.cpp]: still uint32_t here and on the next line

File: src/celengine/universe.cpp [hunk 11]
```diff
@@ -101,7 +101,7 @@ SolarSystem* Universe::getSolarSystem(const Star* star) const
     if (star == nullptr)
         return nullptr;
 
-    uint32_t starNum = star->getCatalogNumber();
+    uint32_t starNum = star->getIndex();
```
Comment (375gnu) [inline src/celengine/universe.cpp]: uint32_t

File: src/celengine/starname.cpp [hunk 12]
```diff
@@ -106,7 +106,7 @@ StarNameDatabase* StarNameDatabase::readNames(istream& in)
 
     while (!failed)
     {
-        uint32_t catalogNumber = Star::InvalidCatalogNumber;
+        uint32_t catalogNumber = AstroCatalog::InvalidIndex;
```
Comment (375gnu) [inline src/celengine/starname.cpp]: uint32_t

File: src/celengine/starname.cpp [hunk 13]
```diff
@@ -19,7 +19,7 @@ using namespace std;
 uint32_t StarNameDatabase::findCatalogNumberByName(const string& name) const
 {
     uint32_t catalogNumber = getCatalogNumberByName(name);
```
Comment (375gnu) [inline src/celengine/starname.cpp]: uint32_t

File: src/celengine/star.h [hunk 14]
```diff
@@ -307,12 +301,10 @@ class Star : public CatEntry
     inline bool hasCorona() const;
 
     enum : uint32_t {
```
Comment (375gnu) [inline src/celengine/star.h]: uint32_t. and please move { to the next line

File: src/celengine/dsodb.h [hunk 15]
```diff
@@ -79,7 +79,7 @@ class DSODatabase
     DSONameDatabase* namesDB{ nullptr };
     DeepSkyObject**  catalogNumberIndex{ nullptr };
     DSOOctree*       octreeRoot{ nullptr };
-    uint32_t         nextAutoCatalogNumber{ 0xfffffffe };
+    AstroCatalog::IndexNumber         nextAutoCatalogNumber{ 0xfffffffe };
```
Comment (375gnu) [inline src/celengine/dsodb.h]: remove extra spaces plz

File: src/celengine/dsodb.cpp [hunk 16]
```diff
@@ -137,7 +137,7 @@ string DSODatabase::getDSONameList(const DeepSkyObject* const & dso, const unsig
 {
     string dsoNames;
 
-    unsigned int catalogNumber   = dso->getCatalogNumber();
+    unsigned int catalogNumber   = dso->getIndex();
```
Comment (375gnu) [inline src/celengine/dsodb.cpp]: unsigned int

**Comment by hv-work** [comment]
> May I use just "auto" here and in similar situations?

yes, this will be much better. let compiler do the dirty job :)

**Comment by pirogronian** [comment]
Ping @375gnu

**Pull request merged**